File: pwiz_tools/Skyline/Test/Quantification/WeightedRegressionTest.cs
/*
 * Original author: Nicholas Shulman <nicksh .at. u.washington.edu>,
 *                  MacCoss Lab, Department of Genome Sciences, UW
 *
 * Copyright 2015 University of Washington - Seattle, WA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System;
using System.Collections.Generic;
using System.Linq;
using MathNet.Numerics.LinearRegression;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using pwiz.Skyline.Util;
using pwiz.SkylineTestUtil;

namespace pwiz.SkylineTest.Quantification
{
    [TestClass]
    public class WeightedRegressionTest : AbstractUnitTest
    {
        [TestMethod]
        public void CompareIntegerWeights()
        {
            Random random = new Random((int) DateTime.Now.Ticks);
            List<double> xRepeatedValues = new List<double>();
            List<double> yRepeatedValues = new List<double>();
            List<int> weights = new List<int>();
            var sampleTuples = new List<Tuple<double[], double>>();
            for (int i = 0; i < 10; i++)
            {
                int weight = random.Next(1, 10);
                weights.Add(weight);
                double x = random.NextDouble();
                xRepeatedValues.AddRange(Enumerable.Repeat(x, weight));
                double y = random.NextDouble();
                yRepeatedValues.AddRange(Enumerable.Repeat(y, weight));
                sampleTuples.Add(new Tuple<double[], double>(new[]{x}, y));
            }
            const double epsilon = 1E-12;
            var repeatedIntercept = Statistics.Intercept(new Statistics(yRepeatedValues),
                new Statistics(xRepeatedValues));
            var repeatedSlope = Statistics.Slope(new Statistics(yRepeatedValues), new Statistics(xRepeatedValues));
            var repeatedSlopeWithoutIntercept = SlopeWithoutIntercept(yRepeatedValues, xRepeatedValues);

            var weightedRegression = WeightedRegression.Weighted(sampleTuples, weights.Select(w=>(double)w).ToArray(), true);
            Assert.AreEqual(repeatedIntercept, weightedRegression[0], epsilon);
            Assert.AreEqual(repeatedSlope, weightedRegression[1], epsilon);
            var weightedRegressionWithoutIntercept = WeightedRegression.Weighted(sampleTuples,
                weights.Select(w => (double) w).ToArray());
            Assert.AreEqual(repeatedSlopeWithoutIntercept, weightedRegressionWithoutIntercept[0], epsilon);
        }

        private static double SlopeWithoutIntercept(List<double> y, List<double> x)
        {
            double dotProduct = 0;
            double sumOfXSquared = 0;
            for (int i = 0; i < y.Count; i++)
            {
                dotProduct += y[i]*x[i];
                sumOfXSquared += x[i]*x[i];
            }
            return dotProduct/sumOfXSquared;
        }
    }
}

File: pwiz_tools/Skyline/Executables/JavaSkylineAlgorithms/src/test/java/edu/washington/gs/skyline/model/quantification/WeightedRegressionTest.java
/*
 * Original author: Nick Shulman <nicksh .at. u.washington.edu>,
 *                  MacCoss Lab, Department of Genome Sciences, UW
 *
 * Copyright 2016 University of Washington - Seattle, WA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package edu.washington.gs.skyline.model.quantification;

import junit.framework.TestCase;
import org.apache.commons.math3.stat.regression.SimpleRegression;

import java.util.Date;
import java.util.Random;

public class WeightedRegressionTest extends TestCase {
    public void testWeighted() {
        Random random = new Random((int) new Date().getTime());
        SimpleRegression simpleRegressionWithIntercept = new SimpleRegression(true);
        SimpleRegression simpleRegressionWithoutIntercept = new SimpleRegression(false);
        final int nPoints = 10;
        double[][] xValues = new double[nPoints][];
        double[] yValues = new double[nPoints];
        double[] weights = new double[nPoints];
        for (int i = 0; i < nPoints; i++)
        {
            int weight = random.nextInt(10) + 1;
            weights[i] = weight;
            double x = random.nextDouble();
            double y = random.nextDouble();
            xValues[i] = new double[]{x};
            yValues[i] = y;
            for (int w = 0; w < weight; w++) {
                simpleRegressionWithIntercept.addData(x, y);
                simpleRegressionWithoutIntercept.addData(x, y);
            }
        }
        final double epsilon = 1E-12;
        double repeatedIntercept = simpleRegressionWithIntercept.getIntercept();
        double repeatedSlope = simpleRegressionWithIntercept.getSlope();

        double[] weightedRegression = WeightedRegression.weighted(xValues, yValues, weights, true);
        assertEquals(repeatedIntercept, weightedRegression[0], epsilon);
        assertEquals(repeatedSlope, weightedRegression[1], epsilon);
        double[] weightedRegressionWithoutIntercept = WeightedRegression.weighted(xValues, yValues, weights, false);

        double repeatedSlopeWithoutIntercept = simpleRegressionWithoutIntercept.getSlope();
        assertEquals(repeatedSlopeWithoutIntercept, weightedRegressionWithoutIntercept[0], epsilon);

    }
}
